### ISSUE TITLE
ValidationError: Value for url cannot be None

### ISSUE THREAD

**Issue by mcopes73** [conda/conda#8869]
## Observed behavior 
We are running into the following error when running `conda install`. 
```
Traceback (most recent call last):
    File "/tmp/conda_wrapper_3255c18b/install/lib/python2.7/site-packages/conda_build/cli/main_build.py", line 456, in main
      execute(sys.argv[1:])
    File "/tmp/conda_wrapper_3255c18b/install/lib/python2.7/site-packages/conda_build/cli/main_build.py", line 447, in execute
      verify=args.verify, variants=args.variants)
    File "/tmp/conda_wrapper_3255c18b/install/lib/python2.7/site-packages/conda_build/api.py", line 208, in build
      notest=notest, need_source_download=need_source_download, variants=variants)
    File "/tmp/conda_wrapper_3255c18b/install/lib/python2.7/site-packages/conda_build/build.py", line 2311, in build_tree
      notest=notest,
    File "/tmp/conda_wrapper_3255c18b/install/lib/python2.7/site-packages/conda_build/build.py", line 1397, in build
      create_build_envs(m, notest)
    File "/tmp/conda_wrapper_3255c18b/install/lib/python2.7/site-packages/conda_build/build.py", line 1285, in create_build_envs
      is_cross=m.is_cross, is_conda=m.name() == 'conda')
    File "/tmp/conda_wrapper_3255c18b/install/lib/python2.7/site-packages/conda_build/environ.py", line 865, in create_env
      execute_actions(actions, index)
    File "/tmp/conda_wrapper_3255c18b/install/lib/python2.7/site-packages/conda/common/io.py", line 46, in decorated
      return f(*args, **kwds)
    File "/tmp/conda_wrapper_3255c18b/install/lib/python2.7/site-packages/conda/plan.py", line 390, in execute_actions
      execute_instructions(plan, index, verbose)
    File "/tmp/conda_wrapper_3255c18b/install/lib/python2.7/site-packages/conda/plan.py", line 600, in execute_instructions
      cmd(state, arg)
    File "/tmp/conda_wrapper_3255c18b/install/lib/python2.7/site-packages/conda/instructions.py", line 70, in UNLINKLINKTRANSACTION_CMD
      unlink_link_transaction.execute()
    File "/tmp/conda_wrapper_3255c18b/install/lib/python2.7/site-packages/conda/core/link.py", line 223, in execute
      self.verify()
    File "/tmp/conda_wrapper_3255c18b/install/lib/python2.7/site-packages/conda/common/io.py", line 46, in decorated
      return f(*args, **kwds)
    File "/tmp/conda_wrapper_3255c18b/install/lib/python2.7/site-packages/conda/core/link.py", line 200, in verify
      self.prepare()
    File "/tmp/conda_wrapper_3255c18b/install/lib/python2.7/site-packages/conda/core/link.py", line 192, in prepare
      stp.remove_specs, stp.update_specs)
    File "/tmp/conda_wrapper_3255c18b/install/lib/python2.7/site-packages/conda/core/link.py", line 267, in _prepare
      for prec, pcrec in zip(link_precs, pkg_cache_recs_to_link))
    File "/tmp/conda_wrapper_3255c18b/install/lib/python2.7/site-packages/conda/core/link.py", line 267, in <genexpr>
      for prec, pcrec in zip(link_precs, pkg_cache_recs_to_link))
    File "/tmp/conda_wrapper_3255c18b/install/lib/python2.7/site-packages/conda/gateways/disk/read.py", line 106, in read_package_info
      paths_data=paths_data,
    File "/tmp/conda_wrapper_3255c18b/install/lib/python2.7/site-packages/conda/_vendor/auxlib/entity.py", line 719, in __call__
      instance = super(EntityType, cls).__call__(*args, **kwargs)
    File "/tmp/conda_wrapper_3255c18b/install/lib/python2.7/site-packages/conda/_vendor/auxlib/entity.py", line 736, in __init__
      setattr(self, key, kwargs[key])
    File "/tmp/conda_wrapper_3255c18b/install/lib/python2.7/site-packages/conda/_vendor/auxlib/entity.py", line 859, in __setattr__
      super(ImmutableEntity, self).__setattr__(attribute, value)
    File "/tmp/conda_wrapper_3255c18b/install/lib/python2.7/site-packages/conda/_vendor/auxlib/entity.py", line 427, in __set__
      instance.__dict__[self.name] = self.validate(instance, self.box(instance, instance.__class__, val))
    File "/tmp/conda_wrapper_3255c18b/install/lib/python2.7/site-packages/conda/_vendor/auxlib/entity.py", line 470, in validate
      raise ValidationError(getattr(self, 'name', 'undefined name'), val)
  conda._vendor.auxlib.exceptions.ValidationError: Value for url cannot be None.
```

## Root cause
We've root caused this to be a `repodata_record.json` file corrpution issue, where our `repodata_record.json` file for one of the packages looks like this:

```
{
  "arch": "x86_64",
  "build": "hdf63c60_0",
  "build_number": 0,
  "channel": "<unknown>",
  "constrains": [],
  "depends": [],
  "features": "",
  "fn": "libgcc-ng-9.1.0-hdf63c60_0",
  "license": "GPL",
  "md5": "238f19c803a6ba7bd6dbd6989e03cbf1",
  "name": "libgcc-ng",
  "platform": "linux",
  "subdir": "linux-64",
  "timestamp": 1560112207081,
  "track_features": "",
  "version": "9.1.0"
}
```
So when we try to read the package record the url winds up being null.

## How to reproduce?

It is possible to get Conda to corrput `repodata_record.json` by having an unextracted tarball in `conda_pkgs_dir` such that it's url is not present in `urls.txt`. If we take a working cache and `rm -rf` the untarred package and then remove it's line from `urls.txt`, the next time we run `conda install` it will corrput the record in this way.

## How can this happen?

This can happen due to a race condition or due to a previous `conda install` having been stopped after downloading a tarball but before writing to `urls.txt`. For example, I have been able to reproduce this situation by exiting an execution in this line: https://github.com/conda/conda/blob/ee6701bab40411afa35a935044b95d769d59216b/conda/core/path_actions.py#L1198 i.e. after downloading the tarball but before writing the url to cache. Note that it can also happen if multiple conda installations are sharing the same cache: one may attempt to extract a package that has just been downloaded but has not been yet written to `urls.txt`.

## Fix

One fix that has worked for us locally has been to check whether we have a url for the tarball or not before extracting. See https://github.com/conda/conda/pull/8868

**Comment by msarahan** [comment]
Awesome, thanks for the thorough analysis and fix!

**Comment by msarahan** [comment]
Fixed by #8868

**Comment by mcopes73** [comment]
Thanks for the quick response!
